### PR TITLE
Gracefully handle partial doc replication

### PR DESCRIPTION
**Pull request by garethbowen**
When saving with bulkDocs it's possible that the docs will be
replicated to the server one at a time. This causes sentinel
hydration to not find the parent doc which then is set to null
and the doc is later persisted.

This change handles this situation gracefully by

- changing sentinel hydration to leave the stub if no doc found
  so data is not lost, and
- changing the contact save service to save the parent doc before
  any children docs so they are replicated first.

This also modifies the enketo service to use bulkDocs instead of
multiple individual puts.

medic/medic-webapp#4200

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

### CODE REVIEW

File: sentinel/lib/lineage.js [hunk 1]
```diff
@@ -78,9 +78,17 @@ const buildHydratedDoc = (doc, lineage) => {
...
+  while (true) {
+    const parent = lineage.shift() || stub;
+    if (!parent) {
+      break;
+    }
+    current.parent = parent;
     current = current.parent;
+    if (stub) {
+      stub = stub.parent;
+    }
```
Comment (SCdF) [inline sentinel/lib/lineage.js]: I'm confused. I feel like this code needs a comment essay. Or <grumble about types>.

What is `doc` vs. `current`? Shouldn't stub be `current.parent`? What situation would `stub` be falsey but `parent` not be? Are stubs being walked up at the same time as parents are being walked up?

File: static/js/services/enketo.js [hunk 2]
```diff
@@ -432,14 +432,21 @@ angular.module('inboxServices').service('Enketo',
     };
 
     var saveDocs = function(docs) {
-      return $q.all(docs.map(function(doc) {
-        return DB()
-          .put(doc)
-          .then(function(res) {
-            doc._rev = res.rev;
-            return doc;
+      return DB().bulkDocs(docs)
```
Comment (SCdF) [inline static/js/services/enketo.js]: So to be clear, this *isn't* a save where order is necessary?

File: sentinel/lib/lineage.js (hunk 1 above)
Comment (garethbowen) [reply sentinel/lib/lineage.js]: > What is doc vs. current?

Doc is whatever was passed in. Current is the parent as you walk up the tree.

> Shouldn't stub be current.parent?

I don't follow this.

> What situation would stub be falsey but parent not be?

I think that was just me being overly defensive. I'll remove the check.

> Are stubs being walked up at the same time as parents are being walked up?

Yes. I'll rename some vars and add some comments.

Comment (SCdF) [reply sentinel/lib/lineage.js]: So when entering the while loop `current` either equals `doc` (line 76) or it equals `doc.contact`. 

In the loop we appear to be shifting parents onto `current.parent`. Which me implies the lineage is `doc[.contact?].parent.parent.parent`

Why would the stub then start from `doc.parent`? Shouldn't it follow whatever we're shifting?

File: static/js/services/enketo.js (hunk 2 above)
Comment (garethbowen) [reply static/js/services/enketo.js]: The order is already correct. The main doc is first in the array, and all the other docs are later.

File: sentinel/lib/lineage.js (hunk 1 above)
Comment (garethbowen) [reply sentinel/lib/lineage.js]: Good spotting. When I made the readability changes a unit test started failing on exactly this. Fixed in the latest commit!

**Pull request merged**